core/src/main/java/org/apache/calcite/sql/validate/AggVisitor.java

```diff
@@ -44,6 +44,9 @@
    * or group auxiliary functions (e.g. {@code TUMBLE_START}). */
   protected final boolean group;
   protected final SqlNameMatcher nameMatcher;
+  /** Whether to find grouping functions (i.e., {@code GROUP_ID},
+   * {@code GROUPING_ID}, {@code GROUPING}).*/
+  private boolean grouping = false;
```

Comment:
Does the `group` variable serve with the same purpose?

Reply:
I don't think the `grouping` field is necessary. If `AggFinder` treats grouping functions (e.g. GROUP_ID) as aggregate functions (e.g. SUM) by default, we get what we need.

Reply:
@hsyuan It is not the same purpose. The `group` variable is used for group functions like `TUMBLE_START`.

Reply:
How about `aggregate`?

Reply:
@julianhyde 
I noticed that AggVisitor treats grouping functions differently as other aggregate functions (`!(operator instanceof SqlAbstractGroupFunction)`).
It is implemented in [CALCITE-2135](https://issues.apache.org/jira/browse/CALCITE-2135).
By default, it ignores these grouping functions. Meanwhile, Calcite validator also deals with the logic specifically in some places. 
For example,
```
sql("select ^group_id()^ from emp")
```
throws `"GROUP_ID operator may only occur in an aggregate query"`

Reply:
By the way, do you think it is a good way to unify them?

Reply:
@hsyuan Yes, the `aggregate` variable determines to find regular aggregate functions. But in `visit(SqlCall call)`, the grouping functions are excluded.

Reply:
@DonnyZone  Thanks for explaining!

Reply:
I agree, we want ```sql("select ^group_id()^ from emp")``` to continue to give an error. Whereas the existence of an aggregate function, say ```sql("select count(*) from emp") would cause the query to be regarded as an aggregate query. So, yes, we will need a boolean field to make `AggFinder` treat aggregate functions (e.g. `COUNT`) and  grouping functions (e.g.`GROUP_ID`) differently when called from `SqlValidatorImpl.isAggregate`.

Are there any other cases where we need to treat them differently? Hopefully not.

Reply:
There are about two cases:
**pattern-1(SqlValidatorTest#testGroupId):** 
`sql("select ^group_id()^ from emp")`
**Expected:** "GROUP_ID operator may only occur in an aggregate query"
**Actual:** OK

**pattern-2(SqlValidatorTest#testGroupId):** 
`sql("select deptno from emp order by ^group_id(deptno)^")`
**Expected:** "GROUP_ID operator may only occur in SELECT, HAVING or ORDER BY clause"
**Actual:** "Aggregate expression is illegal in ORDER BY clause of non-aggregating SELECT"

Reply:
The 2nd Actual is fine. But can you add a test
```
sql("select count(*) from emp order by ^group_id()^")
```
and that should give the error "GROUP_ID operator may only occur in SELECT, HAVING or ORDER BY clause"

Reply:
@julianhyde I tested with /without the patch, it does not throw any exception ~

Reply:
Yes, this patch only enables `AggVisitor` to find grouping functions in Filter, without changing the beheavior in other places.
Without this patch, I also tested remove the condition `!(operator instanceof SqlAbstractGroupFunction)` in `AggVisitor`,  the query does not throw any exceptions.

Reply:
It should throw exception, i think we also need to fix the IN sub-clause just like the filter.